[RFC] Windows : Implement os_get_hostname()

raacker opened the pull request:
TODO : Add codes for Windows in os_get_hostname() env.c
- Call gethostname() function at Windows part
- libuv has uv_getnameinfo function but it uses socket, so It will depend on which ip address given. So I had to try on other Window APIs, and GetHostNameW and gethostname were solution. But GetHostNameW was not supported with WinSock2.h in some computer environment. (You can find it at [GetHostNameW()](https://msdn.microsoft.com/en-us/library/windows/desktop/dn793576%28v=vs.85%29.aspx)). 

But Windows still uses utf-16, so I added utf16_to_utf8 function to it. 

And I did some many silly things (like commit too much.. I just learned --amend). I'll clear that!
Any advice welcome :)

equalsraf commented on src/nvim/os/env.c:
```diff
@@ -26,6 +26,11 @@
 #include <sys/utsname.h>
 #endif
 
+#if defined(__WIN32__) || defined(_WIN32) || defined(WIN32)
```
`#ifdef WIN32` should be enough

raacker replied:
Thank you. I'll change it

equalsraf commented:
Is this only for os_get_hostname? Then e don't need a full replacement for utsname, we can use gethostname() for that.

raacker commented:
That's correct. just call gethostname() and add it to hostname. 

My thought was the source is not mine, but from other developer!

I think it is better to change the files.

equalsraf commented:
If the goal is to fix os_get_hostname, can you change the title to `Windows: implement os_get_hostname()`

I did a search, and there are at least three ways to get the hostname in Windows

1. gethostname()
1. getenv("COMPUTERNAME")
1. GetComputerNameW(), this is what [Vim uses](https://github.com/vim/vim/blob/b9c31e71f5a4653a27e81c21226039bfa35b9131/src/os_win32.c#L3000)

I'm not sure how correct these are with regards to the encoding of the hostname. I think the best would be to use GetComputerNameW() and convert the string using utf16_to_utf8(), but I may be wrong here.

raacker commented:
Thank you for advising! I'll change the title.

Yup that's right.  There are many ways to get host name but they should be tested first.

gethostname works fine though :)

raacker commented:
I'll check those functions first. Thank you very much :)

justinmk commented:
Please first confirm that libuv doesn't already provide this.

raacker commented:
Thank you @justinmk. I didn't know about libuv! 

I found uv_getnameinfo function and uv_getnameinfo_t structure which will give hostname information.

I'm working on now... but I have a question! 

It uses getnameinfo() and it requires socket! So I just added localhost address on it for initial test. But... I think it needs to get actual ip address. isn't it?

raacker committed:
Use GetComputerName to os_get_hostname function
- libuv doens't support such api
- Use WCHAR Array for unicode strings and use utf16_to_utf8 for it

raacker commented:
I'm really sorry about what I'm doing.. Messed up.. 

I think I should get used to with git more.

fwalch commented:
@raacker No need to apologize; it is completely fine if you force-push/update your pull request branch while you are working on it.

raacker commented:
@fwalch  Thank you very much :)

equalsraf commented:
It seems to me that getnameinfo is not what we want here. For ip sockes it will use DNS to determine the hosname, which returns a FQDN, and this might not be what the user is expecting at all. For example for public ip address socket it might return an ISP dependant hostname, or one of those name.local hostnames from your home router

In windows uv_get_nameinfo uses [GetNameInfoW](https://msdn.microsoft.com/en-us/library/windows/desktop/ms738531(v=vs.85).aspx)

raacker commented:
Oh.. It wasn't working well and I try it on test code but that getnameinfo() function was not for machine local hostname! Then I think libuv doens't support this way. I should use windows dependency function! Thank you for your consideration @equalsraf .

raacker commented:
At first, It was not able to find GetHostNameW function in WinSock2.h ( It uses WCHAR so it will be able to get correct hostname in every languages). But I was trying to find GetHostNameW in other computers, and it was surprise.. Seems some windows have GetHostNameW and other don't. I think I should check why this is happening. Sorry I will change it to WIP

jamessan commented:
@equalsraf @justinmk Is there anything else needed to merge this?

justinmk commented:
Needs a test. Create a new file, `test/functional/eval/hostname_spec.lua`

justinmk commented:
I'll post a PR in a few minutes.

justinmk commented on src/nvim/os/env.c:
```diff
@@ -127,11 +131,19 @@ void os_get_hostname(char *hostname, size_t len)
+    *hostname = '\0';
+  } else {
+    int conversion_result = utf16_to_utf8(hostname_unicode,
+                                          &hostname);
```
This doesn't work because `hostname` is reassigned by `utf16_to_utf8`, so the original pointer passed to `os_get_hostname()` isn't updated. Instead a temporary pointer should be passed to `utf16_to_utf8` then copied to `hostname`.
